engine/src/main/java/org/apache/apex/engine/plugin/DefaultApexPluginDispatcher.java

```diff
@@ -0,0 +1,127 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.engine.plugin;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.engine.api.DAGExecutionPluginContext.Handler;
+import org.apache.apex.engine.api.DAGExecutionPluginContext.RegistrationType;
+import org.apache.apex.engine.api.PluginLocator;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.common.util.NameableThreadFactory;
+import com.datatorrent.stram.StramAppContext;
+import com.datatorrent.stram.StreamingContainerManager;
+import com.datatorrent.stram.webapp.AppInfo;
+
+/**
+ * Handle dispatching of events from Stram to Plugins. This implementation creates an executor
+ * service to process the event asynchronously. A separate task {@link DefaultApexPluginDispatcher.ProcessEventTask}
+ * is created to process an event and then submitted to the executor for execution.
+ */
+public class DefaultApexPluginDispatcher extends ApexPluginManager implements ApexPluginDispatcher
+{
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultApexPluginDispatcher.class);
+  private int qsize = 4098;
+  private ArrayBlockingQueue<Runnable> blockingQueue;
+  private ExecutorService executorService;
+  private static final int NUM_MAX_THREADS = 10;
+
+  public DefaultApexPluginDispatcher(PluginLocator locator, StramAppContext context, StreamingContainerManager dmgr, AppInfo.AppStats stats)
+  {
+    super(locator, context, dmgr, stats);
+  }
+
+  @Override
+  public <T> void dispatch(RegistrationType<T> registrationType, T data)
+  {
+    if (executorService != null) {
+      executorService.submit(new ProcessEventTask<>(registrationType, data));
+    }
+  }
+
+  @Override
+  protected void serviceInit(Configuration conf) throws Exception
+  {
+    super.serviceInit(conf);
+    LOG.debug("Creating plugin dispatch queue with size {}", qsize);
+    blockingQueue = new ArrayBlockingQueue<>(qsize);
+    RejectedExecutionHandler handler = new RejectedExecutionHandler()
+    {
+      @Override
+      public void rejectedExecution(Runnable r, ThreadPoolExecutor executor)
+      {
+        try {
+          blockingQueue.remove();
+          executor.submit(r);
+        } catch (NoSuchElementException ex) {
+          // Ignore no-such element as queue may finish, while this handler is called.
+        }
+      }
+    };
+
+    executorService = new ThreadPoolExecutor(1, NUM_MAX_THREADS, 60L, TimeUnit.SECONDS,
```

Comment:
I think the default behavior here should keep things simple for the developer. This currently means the plugin writer has to worry about thread safety. How about going with a single thread executor. You can provide a more advanced version of per plugin single thread executor later on.